internal/p2p/channel_test.go

```diff
@@ -0,0 +1,183 @@
+			},
+		},
+		{
+			Name: "SendWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+				require.Error(t, ch.Send(cctx, Envelope{From: "kip", To: "merlin"}))
+			},
+		},
+		{
+			Name: "SendErrorWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				require.Error(t, ch.SendError(cctx, PeerError{NodeID: "kip", Err: errors.New("merlin")}))
+			},
+		},
+		{
+			Name: "ReceiveEmptyIteratorBlocks",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(1)
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+				out := make(chan bool)
+				go func() {
+					defer close(out)
+					select {
+					case <-ctx.Done():
+					case out <- iter.Next(ctx):
+					}
+				}()
+				select {
+				case <-time.After(10 * time.Millisecond):
+				case <-out:
+					require.Fail(t, "iterator should not advance")
+				}
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "ReceiveWithData",
+			Case: func(ctx context.Context, t *testing.T) {
+				ins, ch := testChannel(1)
+				ins.In <- Envelope{From: "kip", To: "merlin"}
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+				require.True(t, iter.Next(ctx))
+
+				res := iter.Envelope()
+				require.EqualValues(t, "kip", res.From)
+				require.EqualValues(t, "merlin", res.To)
+			},
+		},
+		{
+			Name: "ReceiveWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				iter := ch.Receive(cctx)
+				require.NotNil(t, iter)
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "IteratorWithCanceledContext",
+			Case: func(ctx context.Context, t *testing.T) {
+				_, ch := testChannel(0)
+
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "IteratorCanceledAfterFirstUseBecomesNil",
+			Case: func(ctx context.Context, t *testing.T) {
+				ins, ch := testChannel(1)
+
+				ins.In <- Envelope{From: "kip", To: "merlin"}
+				iter := ch.Receive(ctx)
+				require.NotNil(t, iter)
+
+				require.True(t, iter.Next(ctx))
+
+				res := iter.Envelope()
+				require.EqualValues(t, "kip", res.From)
+				require.EqualValues(t, "merlin", res.To)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+	}
+
+	for idx := range testCases {
+		t.Run(testCases[idx].Name, func(t *testing.T) {
+			t.Cleanup(leaktest.Check(t))
+
+			ctx, cancel := context.WithCancel(bctx)
+			defer cancel()
+
+			testCases[idx].Case(ctx, t)
+		})
+	}
```

Comment:
```suggestion
	for _, tc := range testCases {
		t.Run(tc.Name, func(t *testing.T) {
			t.Cleanup(leaktest.Check(t))

			ctx, cancel := context.WithCancel(bctx)
			defer cancel()

			tc.Case(ctx, t)
		})
	}
```

Reply:
Sure, but that just means it's confusing now, and we'd have to do other things to make these tests parallelizable anyway. That said, I don't object, it just seemed unnecessary 🙂